server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -236,30 +247,78 @@ public boolean downloadTranslog(String primaryTerm, String generation, Path loca
+            try {
+                assert metadata != null && !metadata.isEmpty() && metadata.containsKey(CHECKPOINT_FILE_DATA_KEY);
+                recoverCkpFileUsingMetadata(metadata, location, generation, translogFilename);
+            } catch (Exception e) {
+                throw new IOException("Failed to recover checkpoint file from remote", e);
+            }
+        }
         return true;
     }
 
-    private void downloadToFS(String fileName, Path location, String primaryTerm) throws IOException {
+    /**
+     * Process the provided metadata and tries to recover translog.ckp file to the FS.
+     */
+    private void recoverCkpFileUsingMetadata(Map<String, String> metadata, Path location, String generation, String fileName)
+        throws IOException {
+
+        String ckpFileName = Translog.getCommitCheckpointFileName(Long.parseLong(generation));
+        Path filePath = location.resolve(ckpFileName);
+        // Here, we always override the existing file if present.
+        deleteFileIfExists(filePath);
+
+        String ckpDataBase64 = metadata.get(CHECKPOINT_FILE_DATA_KEY);
+        if (ckpDataBase64 == null) {
+            logger.error("Error processing metadata for translog file: {}", fileName);
+            throw new IllegalStateException(
+                "Checkpoint file data key " + CHECKPOINT_FILE_DATA_KEY + " is expected but not found in metadata for file: " + fileName
+            );
+        }
+        byte[] ckpFileBytes = Base64.getDecoder().decode(ckpDataBase64);
+        Files.write(filePath, ckpFileBytes);
+    }
+
+    private Map<String, String> downloadToFS(String fileName, Path location, String primaryTerm, boolean withMetadata) throws IOException {
         Path filePath = location.resolve(fileName);
         // Here, we always override the existing file if present.
         // We need to change this logic when we introduce incremental download
-        if (Files.exists(filePath)) {
-            Files.delete(filePath);
-        }
+        deleteFileIfExists(filePath);
 
+        Map<String, String> metadata = null;
         boolean downloadStatus = false;
         long bytesToRead = 0, downloadStartTime = System.nanoTime();
-        try (InputStream inputStream = transferService.downloadBlob(remoteDataTransferPath.add(primaryTerm), fileName)) {
-            // Capture number of bytes for stats before reading
-            bytesToRead = inputStream.available();
-            Files.copy(inputStream, filePath);
-            downloadStatus = true;
+        try {
+            if (withMetadata) {
+                try (
+                    FetchBlobResult fetchBlobResult = transferService.downloadBlobWithMetadata(
+                        remoteDataTransferPath.add(primaryTerm),
+                        fileName
+                    )
+                ) {
+                    InputStream inputStream = fetchBlobResult.getInputStream();
```

Comment:
I tried this but this won't work with try with resource stmts. The try-with-resources statement expects a resource declaration otherwise we would need to catch exception and handle the close the inputStream. Not going forward with this change.